components/camera/rtsp/rtsp.go

```diff
@@ -0,0 +1,168 @@
+	Address          string                             `json:"rtsp_address"`
+	IntrinsicParams  *transform.PinholeCameraIntrinsics `json:"intrinsic_parameters,omitempty"`
+	DistortionParams *transform.BrownConrady            `json:"distortion_parameters,omitempty"`
+}
+
+// rtspCamera contains the rtsp client, and the reader function that fulfills the camera interface.
+type rtspCamera struct {
+	gostream.VideoReader
+	client     *gortsplib.Client
+	cancelFunc context.CancelFunc
+}
+
+// Close closes the camera.
+func (rc *rtspCamera) Close(ctx context.Context) error {
+	clientTerminated := liberrors.ErrClientTerminated{}
+	rc.cancelFunc()
+	err := rc.client.Close()
+	if err != nil && !errors.Is(err, clientTerminated) {
+		return err
+	}
+	return nil
+}
+
+// NewRTSPCamera creates a camera client for an RTSP given given the server URL.
+// Right now, only supports servers that have MJPEG video tracks.
+func NewRTSPCamera(ctx context.Context, attrs *Attrs, logger golog.Logger) (camera.Camera, error) {
+	c := &gortsplib.Client{}
+	// parse URL
+	u, err := url.Parse(attrs.Address)
+	if err != nil {
+		return nil, err
+	}
+	// connect to the server - be sure to close it if setup fails.
+	err = c.Start(u.Scheme, u.Host)
+	if err != nil {
+		return nil, err
+	}
+	// find published tracks
+	tracks, baseURL, _, err := c.Describe(u)
+	if err != nil {
+		return nil, multierr.Combine(err, c.Close())
+	}
+
+	// find the MJPEG track
+	var forma *format.MJPEG
+	media := tracks.FindFormat(&forma)
+	if media == nil {
+		err := errors.Errorf("MJPEG track not found in rtsp camera %s", u)
+		return nil, multierr.Combine(err, c.Close())
+	}
+	// get the RTP->MJPEG decoder
+	rtpDec := forma.CreateDecoder()
+
+	// Setup the MJPEG track
+	_, err = c.Setup(media, baseURL, 0, 0)
+	if err != nil {
+		return nil, multierr.Combine(err, c.Close())
+	}
+	// On packet retreival, turn it into an image, and store it in shared memory
+	var latestFrame atomic.Value
+	var gotFirstFrameOnce bool
+	gotFirstFrame := make(chan struct{})
+	c.OnPacketRTP(media, forma, func(pkt *rtp.Packet) {
+		// convert RTP packets into NALUs
+		encoded, _, err := rtpDec.Decode(pkt)
+		if err != nil {
+			return
+		}
+		// convert JPEG images to image.Image
+		img, err := jpeg.Decode(bytes.NewReader(encoded))
+		if err != nil {
+			panic(err)
+		}
+
+		// wait for a frame
+		if img == nil {
+			return
+		}
+		latestFrame.Store(img)
+		if !gotFirstFrameOnce {
+			close(gotFirstFrame)
+			gotFirstFrameOnce = true
+		}
+	})
+	// play the track
+	_, err = c.Play(nil)
+	if err != nil {
+		return nil, multierr.Combine(err, c.Close())
+	}
+	// read the image from shared memory when it is requested
+	cancelCtx, cancel := context.WithCancel(context.Background())
+	reader := gostream.VideoReaderFunc(func(ctx context.Context) (image.Image, func(), error) {
+		select {
+		case <-cancelCtx.Done():
+			return nil, nil, cancelCtx.Err()
+		case <-ctx.Done():
+			return nil, nil, ctx.Err()
+		case <-gotFirstFrame:
```

Comment:
This case will almost always fire and not respect the cancelation. You should do the usual independent check against these two contexts above to prevent this

Reply:
replaced the `case` with `default:` and adding the blocking `<-gotFirstFrame` below